### PR TITLE
Reverse 32bit integer

### LINKED ISSUE #18: Reverse 32bit Integer number
# Reverse Integer

Given a signed 32-bit integer `x`, return `x` with its digits reversed. If reversing `x` causes the value to go outside the signed 32-bit integer range `[-2^31, 2^31 - 1]`, then return `0`. Assume the environment does not allow you to store 64-bit integers (signed or unsigned).

## Examples

- **Example 1:**
  - **Input:** `x = 123`
  - **Output:** `321`

- **Example 2:**
  - **Input:** `x = -123`
  - **Output:** `-321`

- **Example 3:**
  - **Input:** `x = 120`
  - **Output:** `21`

## Constraints:
- `-2^31 <= x <= 2^31 - 1`

### PR DESCRIPTION
**Pull request by SiSaR-Pal**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by SiSaR-Pal** [comment]
This pull request is linked to issue #18

**Comment by Nada-saad635** [comment]
working on review it

File: solutions/reverse_integer.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+def reverse(self, x: int) -> int:
```
Comment (Nada-saad635) [inline solutions/reverse_integer.py]: here i dont thing there is any need for self , what you thing ? why you put it ?

File: solutions/reverse_integer.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    >>> reverse(None, 123)
+    321
+    >>> reverse(None, -123)
+    -321
+    >>> reverse(None, 120)
+    21
+    >>> reverse(None, 0)
+    0
+    """
+    # Define the range for signed 32-bit integers
```
Comment (Nada-saad635) [inline solutions/reverse_integer.py:36]: i think you forget the assertion

File: solutions/reverse_integer.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Restore the sign if the number was negative
+    if negative:
+        reversed_x = -reversed_x
+
+    # Check for overflow
+    if reversed_x < INT_MIN or reversed_x > INT_MAX:
+        return 0
+
+    return reversed_x
```
Comment (Nada-saad635) [inline solutions/reverse_integer.py]: the rest of code seems good

File: solutions/tests/test_reverse_integer.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        self.assertEqual(reverse(None, 1000000003), 0)
+
+    def test_large_negative_number(self):
+        """Test reversing a large negative number that exceeds the 32-bit range."""
+        self.assertEqual(reverse(None, -1000000003), 0)
+
+    def test_palindrome_number(self):
+        """Test reversing a palindrome number to ensure it remains the same."""
+        self.assertEqual(reverse(None, 1221), 1221)
+
```
Comment (Nada-saad635) [inline solutions/tests/test_reverse_integer.py:72]: add    assertion  
example
def test_none_input(self):
        """   """
        with self.assertRaises(AssertionError):
            reverse(None)

File: solutions/reverse_integer.py (hunk 1 above)
Comment (Nada-saad635) [reply solutions/reverse_integer.py]: Variable names help to understand the strategy

Comment (SiSaR-Pal) [reply solutions/reverse_integer.py]: The use of self in the function  indicates that the function is defined as a method of a class. but for this code its not required , I will remove it, You are right its a redundancy

Comment (Nada-saad635) [reply solutions/reverse_integer.py]: for x rewrite for more distractive name

File: solutions/reverse_integer.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    INT_MIN, INT_MAX = -(2**31), 2**31 - 1
+
+    # Determine if the number is negative
+    negative = x < 0
+
+    # Work with the absolute value of x
+    x = abs(x)
+
+    # Reverse the digits using string manipulation
+    reversed_x = int(str(x)[::-1])
```
Comment (Nada-saad635) [inline solutions/reverse_integer.py]: Variables are named with snake_case

Comment (SiSaR-Pal) [reply solutions/reverse_integer.py]: Thanks @Nada-saad635  for your thorough revision
I've modified all the required points

**Review by Nada-saad635 — APPROVED** [review]

**Comment by Nada-saad635** [comment]
good code you can merge it

**Review by AseelOmer — COMMENTED** [review]
Your tests are comprehensive and well-written. Consider adding a test case for very large inputs, e.g., numbers near the 32-bit signed integer limit, to ensure that the overflow detection works correctly

**Review by AseelOmer — APPROVED** [review]
